Fix 'docker cp' mount table explosion, take four

corhere opened the pull request:
* Supersedes #38993
* Supersedes #43583
* Supersedes #44171

**- What I did**
* Fixed #38995
* Fixed #43390
* Made the issue reproducible in a DinD environment so it can be covered by a regression test

**- How I did it**
* I got the issue to reproduce in a DinD container by changing the mount propagation mode to shared, which matches the behaviour of SystemD PID 1 outside of a container.
* I arranged to have the daemon bind-mount the container volumes and perform the archiving operation inside an unshared mount namespace so the mount events would not propagate into the container's namespace.

**- How to verify it**
* Shiny new integration test

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
* Fixed `docker cp` failing with "no space left on device" when the daemon root is mounted into the container as a volume


**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
Add integration test for #38995, #43390

Modify the DinD entrypoint scripts to make the issue reproducible inside
a DinD container.

Co-authored-by: Bjorn Neergaard <bneergaard@mirantis.com>
Signed-off-by: Bjorn Neergaard <bneergaard@mirantis.com>
Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
pkg/archive: create whiteout temp dir under dest

The applyLayer implementation in pkg/chrootarchive has to set the TMPDIR
environment variable so that archive.UnpackLayer() can successfully
create the whiteout-file temp directory. Change UnpackLayer to create
the temporary directory under the destination path so that environment
variables do not need to be touched.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
pkg/chrootarchive: stop reexec'ing before chroot

Unshare the thread's file system attributes and, if applicable, mount
namespace so that the chroot operation does not affect the rest of the
process.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented on pkg/chrootarchive/go_linux.go:
```diff
@@ -0,0 +1,44 @@
+// which fn is executed. Any other goroutines, including ones started from fn,
+// will see the same root directory and file system attributes as the rest of
+// the process.
+func Go(path string, fn func()) error {
```
Short blurb following the discussion in the maintainers meeting; suggestion is to not export this if this function has very specific requirements (w.r.t. spawning goroutines - those goroutines would not be chroot'ed)

corhere committed:
Revert "Use real chroot if daemon is running in a user namespace"

This change was introduced early in the development of rootless support,
before all the kinks were worked out and rootlesskit was built. The
author was testing the daemon by inside a user namespace set up by runc,
observed that the unshare(2) syscall was returning EPERM, and assumed
that it was a fundamental limitation of user namespaces. Seeing as the
kernel documentation (of today) disagrees with that assessment and that
unshare demonstrably works inside user namespaces, I can only assume
that the EPERM was due to a quirk of their test environment, such as a
seccomp filter set up by runc blocking the unshare syscall.
https://github.com/moby/moby/pull/20902#issuecomment-236409406

Mount namespaces are necessary to address #38995 and #43390. Revert the
special-casing so those issues can also be fixed for rootless daemons.

This reverts commit dc950567c105153c0a2f8b40d16b989bbddcdb3c.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
Add reusable chroot and unshare utilities

Refactor pkg/chrootarchive in terms of those utilities.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
daemon: dupe the archive implementation

The Linux implementation needs to diverge significantly from the Windows
one in order to fix platform-specific bugs. Cut the generic
implementation out of daemon/archive.go and paste identical, verbatim
copies of that implementation into daemon/archive_{windows,linux}.go to
make it easier to compare the progression of changes to the respective
implementations through Git history.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
daemon: refactor isOnlineFSOperationPermitted

It is only applicable to Windows so it does not need to be called from
platform-generic code. Fix locking in the Windows implementation.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
daemon: drop Windows-only code from archive_unix.go

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
integration: test more copy edge-cases

The existing archive implementation is not easy to reason about by
reading the source. Prepare to rewrite it by covering more edge cases in
tests. The new test cases were determined by black-box characterizing
the existing behaviour.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
daemon: archive in a dedicated mount namespace

Mounting a container's volumes under its rootfs directory inside the
host mount namespace causes problems with cross-namespace mount
propagation when /var/lib/docker is bind-mounted into the container as a
volume. The mount event propagates into the container's mount namespace,
overmounting the volume, but the propagated unmount events do not fully
reverse the effect. Each archive operation causes the mount table in the
container's mount namespace to grow larger and larger, until the kernel
limiton the number of mounts in a namespace is hit. The only solution to
this issue which is not subject to race conditions or other blocker
caveats is to avoid mounting volumes into the container's rootfs
directory in the host mount namespace in the first place.

Mount the container volumes inside an unshared mount namespace to
prevent any mount events from propagating into any other mount
namespace. Greatly simplify the archiving implementations by also
chrooting into the container rootfs to sidestep the need to resolve
paths in the host.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
container: make path resolution fns Windows-only

The new daemon.containerFSView type covers all the use-cases on Linux
with a much more intuitive API, but is not portable to Windows.
Discourage people from using the old and busted functions in new Linux
code by excluding them entirely from Linux builds.

Signed-off-by: Cory Snider <csnider@mirantis.com>

fuweid commented:
/cc

neersighted reviewed (changes requested):
Overall this is a lot easier to follow than the existing code, even with the use of cancelable asynchronous execution in the `containerFSView` thread. I do wonder if it would make sense to pool/reference count these threads, so that concurrent copy operations do not churn threads; it seems like the only barrier to doing so (which should likely be a follow-up PR) is complexity of implementation.

I tested this PR with both Docker-in-Docker (yay, the `mount --make-rshared` find makes this fully reproducible in dind!) and on a 'bare' Linux system (VM, no Docker-in-Docker), and all works as expected with both the minimal shell-based reproduction from my PR, as well as the stress test based on @kolyshkin's work in #38993.

Some final thoughts:

* Can we drop `container.StatPath` entirely from `container/archive.go`? It seems like the only usages were here.
* There is a moderate cleanup of the path handling code in `archive_unix.go`, but there is some duplication and ambiguity now, comments would help a lot.

neersighted commented:
For posterity, I'm using the following to reproduce the issue:

```bash
docker run --name mount-repro -d -v /var/run:/var/run busybox sleep 1d
docker exec -it mount-repro cat /proc/self/mountinfo | wc -l
docker cp mount-repro:/etc/passwd ./testfile
docker exec -it mount-repro cat /proc/self/mountinfo | wc -l
docker rm -f mount-repro
rm testfile
```

This stress-test hammers the daemon with copies and container restarts and blows up the mount table if the bug is not fixed (and is an improvement of a version of the same created by @kolyshkin's in #38993):

```bash
#!/usr/bin/env bash

set -e

NCONTAINERS=${NCONTAINERS:-$((2*$(nproc)))}
NPARALLEL=${NPARALLEL:-4}
DURATION=${DURATION:-1m}

echo "Creating $NCONTAINERS containers for stress test..."
for _ in $(seq "$NCONTAINERS"); do
  docker run -d -v /var/run:/var/run busybox sleep 1d | tee -a stress-containers
done

OUTDIR=$(mktemp -d)

cleanup() {
  echo "Reaping tasks..."
  for job in $(jobs -p); do
    kill "$job"
  done

  wait

  echo "Killing containers..."
  parallel -u -j"$NPARALLEL" -a stress-containers docker rm -f '{}'
  rm -f stress-containers
  rm -rf "$OUTDIR"
}; trap 'cleanup' EXIT

copy() {
  echo "Starting copy..."
  while test -f stress-containers; do
    parallel -u -j"$NPARALLEL" -a <(shuf stress-containers) docker cp '{}':/etc/passwd "$(mktemp "${OUTDIR}/stress-XXXX")"
  done
}; copy &

restart() {
  echo "Starting restart..."
  while test -f stress-containers; do
    parallel -u -j"$NPARALLEL" -a <(shuf stress-containers) docker restart -t0 '{}'
  done
}; restart &

sleep "$DURATION"
echo "Test complete!"
```

corhere commented on internal/unshare/unshare_linux.go:
```diff
@@ -0,0 +1,174 @@
+// as the rest of the process.
+//
+// The acceptable flags are documented in the [unshare(2)] Linux man-page.
+// The corresponding CLONE_* constants are defined in package [unix].
```
https://go.dev/doc/comment#links

neersighted replied:
Ah, apparently I forgot about the implicit linking semantics 😅

neersighted commented on hack/dind:
```diff
@@ -37,6 +37,10 @@ if [ -f /sys/fs/cgroup/cgroup.controllers ]; then
 
+# Change mount propagation to shared to make the environment more similar to a
+# modern Linux system, e.g. with SystemD as PID 1.
+mount --make-rshared /
```
I think that to the contrary, it is necessary, as it makes dind's environment more like that of a 'real' Linux system (read: modern init system like systemd).

tianon replied:
What I mean is that this change technically changes the way the container shares mounts with the host, right? (It's _not_ restricted just to containers _of_ the Docker-in-Docker instance.)

corhere replied:
No, it _is_ restricted just to containers of the DinD instance. Any mount namespaces cloned from the container's namespace will receive un/mount events from the container's namespace. Mount propagation to/from the host namespace is left unchanged.

corhere commented on internal/unshare/unshare_linux.go:
```diff
@@ -0,0 +1,174 @@
+// including ones started from the function, will see the same execution state
+// as the rest of the process.
+//
+// The acceptable flags are documented in the [unshare(2)] Linux man-page.
```
All flags, including the irreversible ones, are acceptable. There are no user-facing limitations aside from the ones already included in the exported documentation.

neersighted replied:
Ah, right, because we test for reversibility. In that case this looks good -- there's even a mention of `[unix]` for where to find the Go version of the Linux `#define`s.

corhere commented on internal/unshare/unshare_linux.go:
```diff
@@ -0,0 +1,174 @@
+					return
+				}
+				defer func() {
+					if !isReversible || unix.Setns(fd, 0) != nil {
```
Yes it is. I'm really hoping I don't have to repeat lines 114-19 here.

neersighted replied:
Right, that summarizes the overall purpose quite well; however, "if-condition-as-logic" here still is somewhat hidden and without a comment isn't immediately clear.

I'd suggest either a comment or a rewriting of this `if` statement to make the intention clear. I don't think short circuting is so clever that it's a bad idea, so I'd prefer the comment. It wouldn't have to be a whole repeat; instead, something more like:

```
// Short-circuit on the restorability of the thread. If it cannot be restored (due to the original flags or a failure in a iteration), there is no need to keep trying.
```

neersighted replied:
The unfolded version LGTM!

corhere commented on internal/unshare/unshare_linux.go:
```diff
@@ -0,0 +1,174 @@
+// There are more details at https://go.dev/issue/27505.
+//
+// [unshare(2)]: https://man7.org/linux/man-pages/man2/unshare.2.html
+func Go(flags int, setupfn func() error, fn func()) error {
```
Callers could reimplement it themselves with an extra channel and writing `errCh <- err; return` instead of `return err`. It's a bunch of annoying boilerplate which would mirror what is already implemented for the machinery of `Go()` itself.

neersighted replied:
I don't mind it if you think it's useful to most callers -- and it looks like later in the PR you did use it multiple times; LGTM.

corhere commented on daemon/containerfs_linux.go:
```diff
@@ -0,0 +1,221 @@
+	}
+}
+
+// Close waits until any in-flight operations complete and frees all
```
It's more closely associated with the other "core" operations of open and run—their implementations, especially—than the "auxiliary" stat operation which follows.

neersighted replied:
That works for me.

corhere commented on pkg/chrootarchive/archive_unix.go:
```diff
@@ -4,223 +4,63 @@
-	if err := json.NewEncoder(stdin).Encode(options); err != nil {
-		stdin.Close()
-		return nil, errors.Wrap(err, "tar json encode to pipe failed")
+func resolvePath(root, path string) (string, error) {
```
You put in a lot more effort figuring out what this code does than I did. I just spotted some code literally written twice verbatim and DRYed it out. I'd be glad to document it, especially now that the hard part is already done 😁

corhere commented on pkg/chrootarchive/go_linux.go:
```diff
@@ -0,0 +1,92 @@
+	// nondeterministically in the first place. We can accomplish this
+	// simply by locking the main() function to the startup thread! Doing so
+	// excludes any other goroutine from being scheduled on the thread.
+	runtime.LockOSThread()
```
I've looked at the implementation before: `runtime.LockOSThread()` _is_ an incremented lock. You've followed my reasoning exactly.

corhere commented on internal/unshare/unshare_linux.go:
```diff
@@ -0,0 +1,174 @@
+
+	// The following CLONE_NEW* flags are not included because they imply
+	// another, irreversible flag when used with unshare(2).
+	//  - unix.CLONE_NEWIPC:  implies CLONE_SYSVMEM
```
I have already cited my sources: `unshare(2)` and `setns(2)`. Not sure why you're reading the `clone(2)` man page; the semantics of many flags are inverted when passed to `unshare(2)`.

neersighted replied:
Ah, fair enough, I appear to have run out of caffine yesterday:

> This flag has the same effect as the [clone](https://manpages.ubuntu.com/manpages/jammy/en/man2/clone.2.html)(2) CLONE_NEWIPC flag.  Unshare  the  IPC
              namespace,  so  that  the  calling  process has a private copy of the IPC namespace
              which is not shared with any other process.   Specifying  this  flag  automatically
              implies  CLONE_SYSVSEM  as  well.   Use  of CLONE_NEWIPC requires the CAP_SYS_ADMIN
              capability.
LGTM then.

corhere commented on daemon/archive_unix.go:
```diff
@@ -58,66 +48,40 @@ func (daemon *Daemon) containerArchivePath(container *container.Container, path
-	if err != nil {
-		return nil, nil, err
-	}
+	path = archive.PreserveTrailingDotOrSeparator(filepath.Join(string(filepath.Separator), path), path)
```
`path` in the new code is `absPath` in the old code—the absolute path from the container POV. `resolvedPath` is the path from the host POV, which is no longer necessary. Without needing `resolvedPath`, the call to `container.ResolvePath()` can be inlined down to the line above.
https://github.com/moby/moby/blob/542c735926c9208dc0fe9fc02026a110ca0343f9/container/archive.go#L17-L28

The `filepath.Join` call joins `/` with `path`, and [`PreserveTrailingDotOrSeparator` is being used precisely for its documented purpose.](https://pkg.go.dev/github.com/docker/docker/pkg/archive#PreserveTrailingDotOrSeparator) What is there to comment that isn't just repeating the code in prose or duplicating existing docs?

neersighted replied:
Cory and I discussed tweaks offline which were pushed in fixup commits; this now LGTM.

corhere commented on daemon/archive_unix.go:
```diff
@@ -137,77 +101,61 @@ func (daemon *Daemon) containerExtractToDir(container *container.Container, path
-		return errdefs.InvalidParameter(errors.New("extraction point is not a directory"))
-	}
+		// Consider the given path as an absolute path in the container.
+		absPath, err := filepath.EvalSymlinks(filepath.Join(string(filepath.Separator), path))
```
This comment describes the semantics of the `path` value, not the following code which canonicalizes the path.

neersighted replied:
LGTM as above.

corhere commented:
> I do wonder if it would make sense to pool/reference count these threads, so that concurrent copy operations do not churn threads; it seems like the only barrier to doing so (which should likely be a follow-up PR) is complexity of implementation.

I've thought about doing so, too. While it would be fun to write, I think it's a premature optimization until benchmarked otherwise: KISS, YAGNI. This implementation is already orders of magnitude faster than _fork/exec'ing a whole new Go process_ (msecs to µsecs!) so any further optimizations will yield diminishing returns.

> Some final thoughts:
> 
> * Can we drop `container.StatPath` entirely from `container/archive.go`? It seems like the only usages were here.

How can you drop what has already been deleted? :philosoraptor:

Whatever you meant to ask, the answer is probably "no, because Windows support."

> * There is a moderate cleanup of the path handling code in `archive_unix.go`, but there is some duplication and ambiguity now, comments would help a lot.

Which one are you referring to: `daemon/archive_unix.go` or `pkg/chrootarchive/archive_unix.go`?

neersighted reviewed (approved):
Overall looks good to me -- the major increase in readability and clear intention/maintainability of the code is a big win over the previous implementation, even if there is more surface-level complexity (locked goroutines, `unshare(2)`/kernel APIs, cross-goroutine coordination/async).

cpuguy83 commented:
Great work.
Is this going to be back ported to 22.06?

neersighted commented:
It was mentioned at one point, but Cory and Sebastiaan were vehemently opposed due to the late state of the 22.06 branch and the risk/scope of this change.

thaJeztah commented on daemon/containerfs_linux.go:
```diff
@@ -0,0 +1,221 @@
+	"runtime"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
```
This changes made go-multierror a direct dependency (no longer indirect only); looks like CI didn't pick that up (wondering if we could have a way to "fast check" such changes 🤔); opened https://github.com/moby/moby/pull/44453 to fix that.

vijayaramaraju-kalidindi commented:
@neersighted @corhere @thaJeztah - What is the fix for #52201?? Is it addressed in future releases??
